modP2pImpl/src/org/aion/p2p/impl/P2pMgr.java

```diff
@@ -852,16 +625,294 @@ public void shutdown() {
+
+                long startTime = 0, endTime = 0, readEndTime = 0, writeStartTime = 0, writeEndTime = 0;
+
+                if (showLog)
+                    startTime = System.nanoTime();
+
+                if (key.isReadable()) {
+                    P2pMgr.this.read(key);
+
+                    if (showLog)
+                        readEndTime = System.nanoTime();
+                }
+
+                if (key.isWritable()) {
+
+                    if (showLog)
+                        writeStartTime = System.nanoTime();
+
+                    write(channel, key);
+
+                    if (showLog)
+                        writeEndTime = System.nanoTime();
+                }
+
+                if (showLog) {
+                    endTime = System.nanoTime();
+                    // if greater than 1s, something is wrong
+                    if (endTime - startTime > 1000000000) {
+                        System.out.println("channel: "
+                                + channel
+                                + " key-warning total: "
+                                + (endTime - startTime)
+                                + "ns read: " + (readEndTime - startTime)
+                                + "ns write: " + (writeEndTime - writeStartTime) + "ns");
+                        System.out.println("key-stats bb_size: " + byteBuffers.size());
+                    }
+                }
+            } catch (IOException e) {
+                // on any IO exception, cancel the channel, no need to close it should be
+                // closed when channelUnregistered is triggered
+                P2pMgr.this.ioLoop.cancel(key);
+                byteBuffers.clear();
+            }
+        }
+
+        private void write(SelectableChannel channel, SelectionKey key) throws IOException {
+            SocketChannel chan = (SocketChannel) channel;
+            try {
+                LOOP:
+                while (!byteBuffers.isEmpty()) {
+                    ByteBuffer buf = byteBuffers.peek();
+                    // try to write as much as we can
+                    int ret;
+                    while (buf.hasRemaining()) {
+                        ret = chan.write(buf);
+                        if (ret == 0 && buf.hasRemaining()) {
+                            buf.compact();
+                            break LOOP;
+                        }
+                    }
+                    // if we finish processing simply remove
+                    byteBuffers.remove(buf);
+                }
+            } finally {
+                if (byteBuffers.isEmpty())
+                    key.interestOps(SelectionKey.OP_READ);
+            }
+        }
+
+        @Override
+        public void acceptMessage(SelectableChannel channel, SelectionKey key, ByteBuffer buffer) {
+            byteBuffers.add(buffer);
+            key.interestOps(SelectionKey.OP_WRITE);
```

Comment:
Registering ``OP_WRITE`` at initialization causes the selector to spin (99% CPU usage) since the channel is almost always ready for writing see:

http://rox-xmlrpc.sourceforge.net/niotut/

Reply:
:+1:

Reply:
But we're not updating the `interestOps` after finishing writing, right?